### PR TITLE
Feature istio resource renderer

### PR DESCRIPTION
**Pull request by whwalter**
This adds a v1/istio client package to pkg/ingress namespace. The package provides resources for rendering istio v1.beta1 VirtualService and AuthorizationPolicy resources associated with a pre-existing istio Gateway resource and destination service. 

Functions and structs are documented in line. Unit tests are provided for the ConfigureVS and ConfigureAP functions. 

The intent of the IstioClient is to abstract the CRUD for istio resources from the calling controller but that functionality isn't implemented yet.

### CODE REVIEW

File: pkg/ingress/v1/istio/client.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+func (i *IstioClient) UpsertFromConfig(config *IstioConfig) error {
+
+	//TODO: code to create or apply VirtualService and AuthorizationPolicies
+
+	return nil
+}
+
+// IstioConfig contains global configuration for rendering istio VirtualService and AuthorizationPolicy
+// resources from a port to endpoint mapping.
+type IstioConfig struct {
```
Comment (dskatz) [inline pkg/ingress/v1/istio/client.go:49]: nit: consolidate the type definitions in one place.

File: pkg/ingress/v1/istio/client.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+	if ic.ap == nil || ic.vs == nil {
+		return false
+	}
+
+	return true
+}
+
+// ConfigureVS populates the IstioConfig.vs field with a VirtualService associated with the IC.
+// The VS is assocaitged with the gateway assigned to the IstioConfig and maps paths onto the
```
Comment (dskatz) [inline pkg/ingress/v1/istio/client.go]:
```suggestion
// The VS is associated with the gateway assigned to the IstioConfig and maps paths onto the
```

File: pkg/ingress/v1/istio/client.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	vs.Spec.Http = routes
+	ic.vs = &vs
+}
+
+// ConfigureAP configures the IstioConfig.ap field with an AuthorizationPolicy base on the inputs.
+// The AP will contain a single rule that contains the full IP CIDR list and all paths from the
+// endpoint mapping with a glob match. The AP will match the baseURL and baseURL:* hostnames
+// per istio hoat match best practice.
```
Comment (dskatz) [inline pkg/ingress/v1/istio/client.go]:
```suggestion
// per istio host match best practice.
```

File: go.sum [hunk 4]
```diff
@@ -644,6 +644,8 @@ google.golang.org/genproto v0.0.0-20201210142538-e3217bee35cc/go.mod h1:FWY/as6D
 google.golang.org/genproto v0.0.0-20201214200347-8c77b98c765d/go.mod h1:FWY/as6DDZQgahTzZj3fqbO1CbirC29ZNUFHwi0/+no=
 google.golang.org/genproto v0.0.0-20210108203827-ffc7fda8c3d7/go.mod h1:FWY/as6DDZQgahTzZj3fqbO1CbirC29ZNUFHwi0/+no=
 google.golang.org/genproto v0.0.0-20210226172003-ab064af71705/go.mod h1:FWY/as6DDZQgahTzZj3fqbO1CbirC29ZNUFHwi0/+no=
+google.golang.org/genproto v0.0.0-20221227171554-f9683d7f8bef h1:uQ2vjV/sHTsWSqdKeLqmwitzgvjMl7o4IdtHwUDXSJY=
```
Comment (dskatz) [inline go.sum:656]: Needs a `go mod tidy`.  I get a diff for the go.sum locally.

File: pkg/ingress/v1/istio/client.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		gateway:    gw,
+		gwSelector: selector,
+		baseURL:    baseURL,
+	}
+}
+
+// GetVirtualService returns the current configured or unconfigured virtual service
+// if the virtual service is unconfigured it is nil.
+func (ic *IstioConfig) GetVirtualService() *isnetv1beta1.VirtualService {
+	return ic.vs
```
Comment (dskatz) [inline pkg/ingress/v1/istio/client.go]: This should probably return a `DeepCopy` or the `vs` field could be public in the struct.

File: pkg/ingress/v1/istio/client.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+// GetAuthorizationPolicy returns the current configured or unconfigured authorization
+// policy if the authorization policy is unconfigured it is nil.
+func (ic *IstioConfig) GetAuthorizationPolicy() *issecv1beta1.AuthorizationPolicy {
+	return ic.ap
```
Comment (dskatz) [inline pkg/ingress/v1/istio/client.go]: Same comment as on the `GetVirtualService`

**Review by dskatz — DISMISSED** [review]
LGTM.  Minor comments that can be addressed here or in future PRs.

File: pkg/ingress/v1/istio/client.go (hunk 5 above)
Comment (whwalter) [reply pkg/ingress/v1/istio/client.go]: Returning a deep copy is probably the right answer. The vs and ap fields are intentionally private. Outside of this package, the only way to configure the vs or ap is authoritatively the Configure* functions.

**Review by dskatz — DISMISSED** [review]

**Comment by whwalter** [comment]
@dskatz the last commit is just the typo fixes you had suggested.

**Review by dskatz — APPROVED** [review]

**Comment by dskatz** [comment]
> @dskatz the last commit is just the typo fixes you had suggested.

Strange. Apologies.  Didn't think that it would dismiss the review from applying a suggestion.

**Pull request merged**